include/deal.II/lac/dynamic_sparsity_pattern.h

```diff
@@ -1034,20 +1048,50 @@ DynamicSparsityPattern::begin (const size_type r) const
 {
   Assert (r<n_rows(), ExcIndexRangeType<size_type>(r,0,n_rows()));
 
-  // find the first row starting at r that has entries and return the
-  // begin iterator to it. also skip rows for which we do not have
-  // store anything based on the IndexSet given to the sparsity
-  // pattern
-  //
-  // note: row_length(row) returns zero if the row is not locally stored
-  //
-  // TODO: this is way too slow when used in parallel, so do not use it on
-  // non-owned rows
+  if (!have_entries)
+    return iterator(this);
+
+  if (rowset.size() > 0)
+    {
+      // We have an IndexSet that describes the locally owned set. For
+      // performance reasons we need to make sure that we don't do a
+      // linear search over 0..n_rows(). Instead, find the first entry
+      // >= row r in the locally owned set (this is done in log
+      // n_ranges time inside at()). From there, we move forward until
+      // we find a non-empty row. By iterating over the IndexSet instead
+      // of incrementing the row index, we potentially skip over entries
+      // not in the rowset.
+      IndexSet::ElementIterator it = rowset.at(r);
+      if (it == rowset.end())
+        return end(); // we don't own any row between r and the end
+
+      // Instead of using row_length(*it)==0 in the while loop below,
+      // which involves an expensive index_within_set() call, we
+      // look at the lines vector directly. This works, because we are
+      // walking over this vector entry by entry anyways.
+      size_type rowindex = rowset.index_within_set(*it);
+
+      while (it!=rowset.end()
+             && lines[rowindex].entries.size()==0)
+        {
+          ++it;
+          ++rowindex;
+        }
+
+      if (it == rowset.end())
+        return end();
+      else
+        return iterator(this, *it, 0);
+    }
+
+  // Without an index set we have to do a linear search starting at
+  // row r until we find a non-empty one
   size_type row = r;
-  while ((row<n_rows())
-         &&
-         (row_length(row)==0))
-    ++row;
+
+  while (row<n_rows() && row_length(row)==0)
```

Comment:
Would it make sense to check for `lines[row].entries.size()==0` here, given that we know that there is no index set? This way we avoid the indirection `if (rowset.size()==0)`. I'm not sure whether a compiler can deduce that this (trivial) check is safe to remove...